### PR TITLE
fix(proposals): order the bounded list window by CreatedAt, not defer-inflated ExpiresAt (#1247)

### LINKED ISSUE #1247: Defer-inflated ExpiresAt can displace fresher pending proposals out of the list limit window
**Severity:** LOW (surfaced by the #1245 adversarial review). Unreachable in the current single-user/personal-use deployment; tracked for correctness.

## Problem

`AutomationProposalRepository.GetLimitedWithOperationsAsync` (the shared path behind `GetByUserIdAsync` / `GetByBoardIdAsync` / `GetByStatusAsync` / `GetByRiskLevelAsync`) selects the top `boundedLimit` rows via `OrderByDescending(p => p.ExpiresAt)`.

ADR-0042's defer design pushes a snoozed proposal's `ExpiresAt` to `DeferredUntil + 24h grace` (i.e. **>** the default 24h TTL) so it can never silently expire mid-snooze. As a side effect, a once-deferred proposal that has resurfaced carries a **later** `ExpiresAt` than freshly-created proposals and sorts to the very top of the selection window.

If a user's pending set ever exceeds the (over-fetched) limit, a resurfaced deferred proposal occupies a top slot and can push a newer, non-deferred pending proposal **out of the returned window**, hiding it. The defer `WHERE`-filter is correct; this is purely a selection-ordering interaction.

## Why not fixed in #1245

- It only manifests when the pending count exceeds the limit (~hundreds of proposals) — impossible in the current single-user personal-use deployment.
- The clean fix is to make the top-N **selection** order by `CreatedAt` (matching the final display sort), but `CreatedAt` is a `DateTimeOffset`, and EF Core + SQLite cannot reliably `ORDER BY` a `DateTimeOffset` in LINQ (same landmine handled in #1238/#1239 via raw SQL). Bolting a raw-SQL `ORDER BY CreatedAt DESC LIMIT` onto this shared hot path is disproportionate for a LOW that the use case can't hit.

## Proposed fix

Push a `CreatedAt`-desc bounded selection into SQL (raw-SQL `ORDER BY CreatedAt DESC LIMIT` + in-memory re-materialize with `Include(Operations)`), mirroring the pattern in #1238. Add a test asserting a resurfaced deferred proposal does not evict an unrelated fresher pending proposal from a full page.

## References
- `backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs` (`GetLimitedWithOperationsAsync`)
- ADR-0042 (defer pushes `ExpiresAt`); #1238/#1239 (SQLite `DateTimeOffset` raw-SQL ordering precedent)
- Found in PR #1245 review.

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Fixes #1247. `AutomationProposalRepository.GetLimitedWithOperationsAsync` (the shared path behind `GetByUserIdAsync` / `GetByBoardIdAsync` / `GetByStatusAsync` / `GetByRiskLevelAsync`) selected the top-N proposal IDs with `OrderByDescending(p => p.ExpiresAt)`.

ADR-0042's defer design pushes a snoozed proposal's `ExpiresAt` to `DeferredUntil + 24h grace` (greater than the normal TTL) so it can't silently expire mid-snooze. Once such a proposal **resurfaces** (`DeferredUntil <= now`, passing the visibility filter), it still carries that inflated `ExpiresAt` and sorts to the top of the selection window — so on a full page it can evict a fresher, non-deferred pending proposal, hiding it.

## Fix

Select the freshest N by **`CreatedAt`** — the same order as the method's final display sort — so the bounded window is consistent and a resurfaced deferred proposal can't displace a newer one. `CreatedAt` is a `DateTimeOffset`, which EF Core + SQLite can't `ORDER BY` in LINQ (the landmine handled via raw SQL in #1238/#1239; `ExpiresAt` is a `DateTime`, which is why the old ordering translated), so the fix fetches the lightweight `(Id, CreatedAt)` projection and orders it in memory. The over-fetch is two columns and negligible at single-user scale; the heavy `Include(Operations)` read stays bounded to the selected N.

## Testing

- New real-SQLite test `GetByUserIdAsync_ResurfacedDeferredProposal_DoesNotEvictFresherPendingFromFullPage`: 3 candidates, `limit: 2` — an old resurfaced deferred proposal (inflated `ExpiresAt`, oldest `CreatedAt`, `DeferredUntil` in the past) must not occupy a top slot; the two freshest-by-`CreatedAt` pending proposals are returned. Fails under the prior `ExpiresAt` ordering, passes under `CreatedAt`.
- Local (Release): `AutomationProposalRepositoryIntegrationTests` 21/21; proposal-related Api.Tests 216/216 and Application.Tests 263/263; backend build clean. (The final display order — already `CreatedAt DESC` — is unchanged; only which items fall in the bounded window changed, now consistently the freshest.)

## Notes

- Severity is LOW (per the issue): only manifests when a user's pending set exceeds the over-fetched limit (~hundreds), which the single-user personal-use deployment can't realistically hit. Fixed for correctness; the change is dependency-free and behavior-preserving for the common (under-limit) case.

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Two independent adversarial reviews — findings + fixes

Ran two independent adversarial reviews (correctness lens + perf/regression lens). Both confirmed the core fix is correct, real, and well-tested (the new test genuinely discriminates: fails under the old `ExpiresAt` ordering, passes under `CreatedAt`). Two findings overlapped; both are now fixed (`d5788e5c`).

| # | Sev | Lens | Finding | Resolution |
|---|-----|------|---------|------------|
| A | MEDIUM (perf) / LOW (correctness) | both | The first cut selected the top-N **in memory** (`Select(Id, CreatedAt).ToList()` then `OrderBy().Take()`), reintroducing an **unbounded read** on the polled per-user `GetByUserIdAsync` path — terminal proposals (Applied/Rejected/Failed/Approved) are never pruned, so they accumulate and grow the over-fetch. Counter to the recent #1195/#1238/#1209 "bound at DB" direction, when an in-repo precedent exists. | **Fixed** `d5788e5c`: rework to push `ORDER BY CreatedAt DESC, Id LIMIT` into the DB via raw SQL on SQLite (mirrors `NotificationRepository`/`AgentRunRepository`), with a strongly-typed LINQ path for non-SQLite. Read is bounded at the DB again, still ordered by `CreatedAt` (not the defer-inflated `ExpiresAt`), so #1247 stays fixed. The filter column is validated against a fixed allowlist before reaching raw SQL (injection-safe); values are bound parameters. |
| B | MEDIUM | both | No secondary sort key — same-`CreatedAt` proposals (created in the same batch) could nondeterministically flicker in/out of a full page (and within-page order). | **Fixed** `d5788e5c`: added `Id` as a deterministic tiebreaker (`ORDER BY CreatedAt DESC, Id`) in both the selection and display sorts, mirroring `NotificationRepository`. New test `GetByUserIdAsync_EqualCreatedAt_SelectsDeterministicallyViaIdTiebreaker` (two reads with tied `CreatedAt` return the same row). |
| C | LOW (pre-existing, out of scope) | both | `ProposalHousekeepingWorker` expires via the bounded `GetByStatusAsync` window, so a >100 pending backlog can leave expired proposals un-expired; it should use the unbounded `GetExpiredAsync`. Not introduced by this PR. | **Tracked** — seeded #1259 (per the review policy: out-of-scope findings get an issue, not a silent drop). |

### Items reviewers checked and cleared
- Filter-before-projection unchanged (candidate set identical; only ordering changed).
- Two-query consistency, IN-clause bound (≤ limit), no N+1, empty/`limit<=0` paths — all preserved.
- No consumer depended on `ExpiresAt`-ordered selection; the display order was already `CreatedAt DESC`. The change is strictly an improvement scoped to the `>limit` edge case.

### Verification after rework (local, Release)
- Backend build clean.
- `AutomationProposalRepositoryIntegrationTests` **22/22** (real SQLite; +the new tie-determinism test).
- Proposal-related Api.Tests **217/217**, Application.Tests **263/263**.
- The SQLite path (production) is fully exercised; the non-SQLite LINQ branch is defensive parity (dormant — no test targets Postgres for these methods, same as before this change). Full suite runs in CI.

File: backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs [hunk 1]
```diff
@@ -355,12 +355,26 @@ private async Task<IReadOnlyList<AutomationProposal>> GetLimitedWithOperationsAs
...
+        // bounded to the selected N.
+        var candidates = await baseQuery
+            .Select(p => new { p.Id, p.CreatedAt })
             .ToListAsync(cancellationToken);
 
+        var topProposalIds = candidates
+            .OrderByDescending(c => c.CreatedAt)
+            .Take(boundedLimit)
+            .Select(c => c.Id)
+            .ToList();
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs]: Fixed in `d5788e5c`: the bounded selection now pushes `ORDER BY CreatedAt DESC, Id LIMIT` into the database via raw SQL on SQLite (StringBuilder + positional bound params; the filter column is validated against a fixed allowlist before interpolation, so it's injection-safe), with a strongly-typed LINQ path for non-SQLite. No more client-side materialization of all candidates. I went with the raw-SQL pattern (the established convention in `NotificationRepository`/`AgentRunRepository`, ADR-0023) rather than a global `DateTimeOffset -> string` `HaveConversion`: the global converter changes the storage/comparison mapping of *every* DateTimeOffset column repo-wide (migrations + comparison semantics everywhere) — a much larger blast radius than this localized read needs.

File: backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs [hunk 2]
```diff
@@ -355,12 +355,26 @@ private async Task<IReadOnlyList<AutomationProposal>> GetLimitedWithOperationsAs
...
+        // DeferredUntil + 24h grace (> the normal TTL), so once it resurfaces (DeferredUntil <= now, passing
+        // the filter above) it still carries an inflated ExpiresAt and sorts above newer proposals (#1247).
+        // CreatedAt is a DateTimeOffset, which EF Core + SQLite cannot ORDER BY in LINQ (the landmine handled
+        // via raw SQL in #1238/#1239; ExpiresAt is a DateTime, which is why the old ordering translated), so
+        // fetch the lightweight (Id, CreatedAt) projection and order in memory. The over-fetch is just two
+        // columns and is negligible at single-user scale; the heavy Include(Operations) read below stays
+        // bounded to the selected N.
+        var candidates = await baseQuery
+            .Select(p => new { p.Id, p.CreatedAt })
             .ToListAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs]: Fixed in `d5788e5c` exactly as suggested: `ORDER BY CreatedAt DESC, Id` + `LIMIT` are now in SQL (SQLite raw-SQL pattern), so the per-user/board/status/risk reads are bounded at the DB again instead of materializing every matching row.

File: backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs [hunk 3]
```diff
@@ -355,12 +355,26 @@ private async Task<IReadOnlyList<AutomationProposal>> GetLimitedWithOperationsAs
...
 
+        var topProposalIds = candidates
+            .OrderByDescending(c => c.CreatedAt)
+            .Take(boundedLimit)
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Infrastructure/Repositories/AutomationProposalRepository.cs]: Good catch — this is the housekeeping-worker expiry sweep, which is independently wrong (a bounded display-order window is the wrong tool for "find everything expired", under both the old ExpiresAt and the new CreatedAt ordering). The correct fix is to have `ProposalHousekeepingWorker` use the purpose-built unbounded `GetExpiredAsync()` (which already filters `ExpiresAt < now`). Both independent adversarial reviews flagged this as pre-existing and out of scope for this ordering PR, so I've tracked it as #1259 rather than expand this PR (it's also gated on >100 pending, unreachable in the single-user deployment). This PR's CreatedAt ordering does not introduce the worker gap; #1259 fixes it at the right layer.

**Comment by Chris0Jeky** [comment]
## Independent adversarial review of the raw-SQL rework — SAFE TO MERGE

Since the rework (raw-SQL DB-bounding) was new code not covered by the earlier reviews (which assessed the in-memory version), I ran a focused security/correctness adversarial review of the final state. Verdict: **SAFE TO MERGE**.

Confirmed clean: **SQL injection** (`filterColumn` is always a hardcoded `nameof(...)` from the 4 callers and validated against a fixed allowlist *before* interpolation; all values are bound `{N}` params); **FromSqlRaw + Include** (the inner `LIMIT` bounds the principal rows before the operations join; in-memory re-sort is sound); **Guid/enum/DateTime parameter binding** (matches EF's column storage — empirically validated, no silent zero-row mismatch); **defer-clause boolean precedence + positional indices** (correct, no off-by-one).

One **LOW** nit fixed (`b557aa64`): the SQLite in-memory re-sort used `.ThenBy(p => p.Id)` (Guid.CompareTo), which disagrees with the raw SQL's `ORDER BY ... Id` (SQLite TEXT comparison). Switched to `.ThenBy(p => p.Id.ToString(), StringComparer.Ordinal)` to match the DB ordering, mirroring `LlmQueueRepository.GetCapturesByUserAsync`. Cosmetic-only (the set is fixed by SQL; these methods don't paginate).

Re-verified: build clean, `AutomationProposalRepositoryIntegrationTests` 22/22.
